### ISSUE TITLE
Assertion failed with Program Fixpoint

### ISSUE THREAD

**Issue by Tuplanolla** [coq/coq#11586]
#### Version

Coq 8.8.2 and 8.11.0

#### Operating system

Red Hat Enterprise Linux Workstation 7.4

#### Description of the problem

The following program causes an anomaly.

```coq
From Coq Require Arith Program.

Inductive vector (A : Type) : nat -> Type :=
  | nil : vector A O
  | cons : forall {n : nat}, A -> vector A n -> vector A (S n).

Arguments vector A n.
Arguments nil {A}.
Arguments cons {A n} x xs.

Import Arith Nat Program.

(** Tactics work. *)

Fixpoint nth {A : Type} {m : nat}
  (xs : vector A m) (n : {i : nat | i < m}) : A.
Proof.
  destruct xs as [| p y ys]; destruct n as [i H].
  - exfalso. apply (nlt_0_r i). apply H.
  - destruct i as [| q].
    + apply y.
    + apply (nth _ _ ys). exists q. apply (lt_S_n _ _). apply H. Defined.

(** Terms work. *)

Fixpoint nth' {A : Type} {m : nat}
  (xs : vector A m) (n : {i : nat | i < m}) : A :=
  match xs, n with
  | @nil _, exist _ i H => False_rect A (nlt_0_r _ H)
  | @cons _ p y ys, exist _ O H => y
  | @cons _ p y ys, exist _ (S q) H => nth' ys (exist _ q (lt_S_n _ _ H))
  end.

(** Program does not work. *)

Program Fixpoint nth'' {A : Type} {m : nat}
  (xs : vector A m) (n : {i : nat | i < m}) : A :=
  match xs, n with
  | nil, exist _ i H => !
  | cons y ys, exist _ O H => y
  | cons y ys, exist _ (S q) H => nth'' ys (exist _ q _)
  end.
```

```
Anomaly "File "pretyping/cases.ml", line 2031, characters 2-8: Assertion failed."
Please report at http://coq.inria.fr/bugs/.
```

#### Notes

This may be a duplicate of #11030,
although this way to reproduce the bug is less contrived and
applies to a wider range of versions.

**Comment by Alizter** [comment]
Still an issue in 8.15+alpha.

**Comment by herbelin** [comment]
Fixed in #18921. Note that in `Program` style, the fixpoint should be without an explicit `exist`:
```coq
Program Fixpoint nth'' {A : Type} {m : nat}
  (xs : vector A m) (n : {i : nat | i < m}) : A :=
  match xs, n with
  | nil, i => !
  | cons y ys, O => y
  | cons y ys, (S q) => nth'' ys (exist _ q _)
  end.
Next Obligation. ... Qed.
Next Obligation. ... Qed.
```
